client2/thin.go

```diff
@@ -0,0 +1,308 @@
+// SPDX-FileCopyrightText: © 2023 David Stainton
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package client2
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"sync"
+
+	"github.com/charmbracelet/log"
+	"github.com/fxamacker/cbor/v2"
+
+	cpki "github.com/katzenpost/katzenpost/core/pki"
+	sConstants "github.com/katzenpost/katzenpost/core/sphinx/constants"
+	"github.com/katzenpost/katzenpost/core/worker"
+)
+
+// ThinResponse is used to encapsulate a message response
+// that are passed to the client application.
+type ThinResponse struct {
+
+	// SURBID, a unique indentifier for this response,
+	// which should precisely match the application's chosen
+	// SURBID of the sent message.
+	SURBID *[sConstants.SURBIDLength]byte
+
+	ID *[MessageIDLength]byte
+
+	// Payload is the decrypted payload plaintext.
+	Payload []byte
+}
+
+// ThinClient is the client that handles communication between the mixnet application
+// and the client daemon. It does not do any encryption or decryption or checking
+// of cryptographic signatures; those responsibilities are left to the client daemon
+// process.
+type ThinClient struct {
+	worker.Worker
+
+	log          *log.Logger
+	unixConn     *net.UnixConn
+	destUnixAddr *net.UnixAddr
+
+	pkidoc      *cpki.Document
+	pkidocMutex sync.RWMutex
+
+	isConnected bool
+
+	receivedCh chan ThinResponse
+}
+
+// NewThinClient creates a new thing client.
+func NewThinClient() *ThinClient {
+	return &ThinClient{
+		log: log.NewWithOptions(os.Stderr, log.Options{
+			Prefix: "thin_client",
+			Level:  log.DebugLevel,
+		}),
+		receivedCh: make(chan ThinResponse),
+	}
+}
+
+// Close halts the thin client worker thread and closes the socket
+// connection with the client daemon.
+func (t *ThinClient) Close() error {
```

Comment:
not sure but should this also close the `receivedCh` to signal it to downstream readers?

Reply:
fixed